src/v2/manifest/mod.rs

```diff
@@ -238,3 +270,87 @@ fn to_mimes(v: &[&str]) -> Result<Vec<mime::Mime>> {
         .collect();
     Ok(res)
 }
+
+// Evaluate the `MediaTypes` from the the request header.
+fn evaluate_media_type(
+    content_type: Option<&http::header::HeaderValue>,
+    url: &Url,
+) -> Result<mediatypes::MediaTypes> {
+    let header_content_type = content_type
+        .map(|hv| hv.to_str())
+        .map(std::result::Result::unwrap_or_default);
+
+    let is_pulp_based = url.path().starts_with("/pulp/docker/v2");
+
+    match (header_content_type, is_pulp_based) {
+        (Some(header_value), false) => {
+            mediatypes::MediaTypes::from_str(header_value).map_err(Into::into)
+        }
+        (None, false) => Err(Error::from(
+            "no header_content_type given and no workaround to apply".to_string(),
+        )),
+        (Some(header_value), true) => {
+            // TODO: remove this workaround once Satellite returns a proper content-type here
+            match header_value {
+                "application/x-troff-man" => {
+                    trace!("Applying workaround for pulp-based registries, e.g. Satellite");
+                    mediatypes::MediaTypes::from_str(
+                        "application/vnd.docker.distribution.manifest.v1+prettyjws",
+                    )
+                    .map_err(Into::into)
+                }
+                _ => {
+                    debug!("Received content-type '{}' from pulp-based registry. Feeling lucky and trying to parse it...", header_value);
+                    mediatypes::MediaTypes::from_str(header_value).map_err(Into::into)
+                }
+            }
+        }
+        (None, true) => {
+            trace!("Applying workaround for pulp-based registries, e.g. Satellite");
+            mediatypes::MediaTypes::from_str(
+                "application/vnd.docker.distribution.manifest.v1+prettyjws",
+            )
+            .map_err(Into::into)
+        }
+    }
+}
+
+/// Umbrella type for common actions on the different manifest schema types
+#[derive(Debug)]
+pub enum Manifest {
+    S1Signed(manifest_schema1::ManifestSchema1Signed),
+    S2(manifest_schema2::ManifestSchema2),
+    ML(manifest_schema2::ManifestList),
+}
+
+impl Manifest {
+    /// List digests of all layers referenced by this manifest, if available.
+    ///
+    /// The returned layers list is ordered starting with the base image first.
+    pub fn layers_digests(&self) -> Result<Vec<String>> {
+        match self {
+            Manifest::S1Signed(m) => Ok(m.get_layers()),
+            Manifest::S2(m) => Ok(m.get_layers()),
+            // Manifest::ML(_) => TODO(steveeJ),
```

Comment:
To get layers out of a manifest-list, you need to chose an architecture so you know which referenced digest to descend into.  If the API is `layers_digests(&self, architecture: Option<String>) -> ...`, then you can do that.  If the option is passed, you want a check for the `S1Signed` and `S2` that looks into the config to see if the incoming architecture matches the value set there.  In the unset-option case, there is only one choice (`S1Signed` and `S2`, but also `ML` if there happens to only be a single architecture).